op-verifier/contracts/OPVerifier.sol

```diff
@@ -1,43 +1,177 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.17;
 
-import { IEVMVerifier } from "@ensdomains/evm-verifier/contracts/IEVMVerifier.sol";
-import { RLPReader } from "@eth-optimism/contracts-bedrock/src/libraries/rlp/RLPReader.sol";
-import { StateProof, EVMProofHelper } from "@ensdomains/evm-verifier/contracts/EVMProofHelper.sol";
-import { Types } from "@eth-optimism/contracts-bedrock/src/libraries/Types.sol";
-import { Hashing } from "@eth-optimism/contracts-bedrock/src/libraries/Hashing.sol";
+import {IEVMVerifier} from '@ensdomains/evm-verifier/contracts/IEVMVerifier.sol';
+import {RLPReader} from '@eth-optimism/contracts-bedrock/src/libraries/rlp/RLPReader.sol';
+import {StateProof, EVMProofHelper} from '@ensdomains/evm-verifier/contracts/EVMProofHelper.sol';
+import {Types} from '@eth-optimism/contracts-bedrock/src/libraries/Types.sol';
+import {Hashing} from '@eth-optimism/contracts-bedrock/src/libraries/Hashing.sol';
+import './interfaces/IDisputeGame.sol';
+import './interfaces/IDisputeGameFactory.sol';
+
+enum OPWitnessProofType {
+    L2OutputOracle,
+    DisputeGame
+}
 
 struct OPWitnessData {
-    uint256 l2OutputIndex;
+    OPWitnessProofType proofType;
+    uint256 index;
     Types.OutputRootProof outputRootProof;
 }
 
 interface IL2OutputOracle {
-    function getL2Output(uint256 _l2OutputIndex) external view returns (Types.OutputProposal memory);
+    function getL2Output(
+        uint256 _l2OutputIndex
+    ) external view returns (Types.OutputProposal memory);
+}
+
+interface IOptimismPortal {
+    function l2Oracle() external view returns (IL2OutputOracle);
```

Comment:
Can you point me in the source code of optimism repo where it's defined? I can't find the function on etherscan's verified code https://sepolia.etherscan.io/address/0x16Fc5058F25648194471939df75CF27A2fdC48BC#readProxyContract nor on their github repo https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L1/OptimismPortal2.sol

Reply:
Old contract and new contract have different method but always deployed to the same address (Upgradeable Proxy)

* OptimismPortal can resolve to l2Oracle only on the old contract
* OptimismPortal can resolve to disputeGameFactory only on the new contract (Because old method is replaced by spacer_54_0_20)

This makes it possible to automatically determine whether OP is using a new or old implementation

For the source code
* Old: https://github.com/ethereum-optimism/optimism/blob/2b3f00df3e865bd55369c5d6bf4a263839b8f0cc/packages/contracts-bedrock/src/L1/OptimismPortal.sol#L71
* New: https://github.com/ethereum-optimism/optimism/blob/2b3f00df3e865bd55369c5d6bf4a263839b8f0cc/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L80

Probably I should use try-catch. Will test on remix first to see what's the behaviour of calling an undefined function.

Reply:
I see. so your code supports pre-post migrations?

Reply:
Yes, for OP Stack chain that hasn't upgraded